dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -269,12 +269,28 @@ NaryOpNode<BinaryOp>::NaryOpNode(Node* node_ptr)
     add_node(node_ptr);
 }
 
+// Enforce that the given span is nonempty and return the first element
+template <class T>
+T& nonempty(std::span<T> node_ptrs) {
+    if (node_ptrs.empty()) {
+        throw std::invalid_argument("Must supply at least one predecessor");
+    }
+    return node_ptrs[0];
+}
+
 template <class BinaryOp>
-NaryOpNode<BinaryOp>::NaryOpNode(std::vector<Node*>& node_ptrs)
-        : Node(),
-          ArrayOutputMixin(dynamic_cast<const Array*>(verify_node_list_(node_ptrs))->shape()) {
-    for (const auto& node_ptr : node_ptrs) {
-        add_node(node_ptr);
+NaryOpNode<BinaryOp>::NaryOpNode(std::span<Node*> node_ptrs)
+        : Node(), ArrayOutputMixin(dynamic_cast<const Array*>(nonempty(node_ptrs))->shape()) {
+    for (Node* ptr : node_ptrs) {
+        add_node(ptr);
+    }
+}
+
+template <class BinaryOp>
+NaryOpNode<BinaryOp>::NaryOpNode(std::span<Array*> array_ptrs)
+        : Node(), ArrayOutputMixin(nonempty(array_ptrs)->shape()) {
+    for (Array* ptr : array_ptrs) {
+        add_node(dynamic_cast<Node*>(ptr));
```

Comment:
Should we add an assertion that the casting doesn't return `nullptr`?

Reply:
We should. Also, turns out `add_node()` casts to `Array*` .... Will clean that up too.

Reply:
Ok, added some checks. We still do two dynamic casts, but IMO the right fix is deeper. So we can/should live with it for now.